File: codewars/5kyu/vector-class/vector-class.py
# vector-class
# https://www.codewars.com/kata/526dad7f8c0eb5c4640000a4


from unittest import TestCase

from math import sqrt


class Vector:
    def __init__(self, values):
        self.val = values

    def compare_length(self, values):
        if len(self.val) != len(values.val):
            raise Exception()

    def add(self, values):
        self.compare_length(values)
        return Vector(list(map(lambda v: sum(v), list(zip(self.val, values.val)))))

    def subtract(self, values):
        self.compare_length(values)
        return Vector(list(map(lambda v: v[0] - v[1], list(zip(self.val, values.val)))))

    def dot(self, values):
        self.compare_length(values)
        return sum(map(lambda v: v[0] * v[1], list(zip(self.val, values.val))))

    def norm(self):
        return sqrt(sum(list(map(lambda v: v ** 2, self.val))))

    def equals(self, values):
        return len(
            list(
                filter(
                    lambda b: b is False, list(
                        map(lambda v: v[0] == v[1], list(zip(self.val, values.val)))
                    )
                )
            )
        ) == 0

    def __str__(self):
        return '(' + ','.join(map(str, self.val)) + ')'


a = Vector([1, 2])
b = Vector([3, 4])

TestCase().assertTrue(a.add(b).equals(Vector([4, 6])))

a = Vector([1, 2, 3])
b = Vector([3, 4, 5])

TestCase().assertTrue(a.add(b).equals(Vector([4, 6, 8])))
TestCase().assertTrue(a.subtract(b).equals(Vector([-2, -2, -2])))
TestCase().assertEqual(a.dot(b), 26)
TestCase().assertEqual(a.norm(), 14 ** 0.5)

File: codewars/5kyu/vector-class/vector-class.js
// vector-class
// https://www.codewars.com/kata/526dad7f8c0eb5c4640000a4

class Vector {
  constructor(initialValues) {
    this.val = initialValues;
  }
  
  compareLength(vector) {
    if (this.val.length !== vector.val.length) {
      throw Error();
    }
  }
  
  add(vector) {
    this.compareLength(vector);
    return new Vector(this.val.map((v, i) => v + vector.val[i]));
  }
  
  subtract(vector) {
    this.compareLength(vector);
    return new Vector(this.val.map((v, i) => v - vector.val[i]));
  }
  
  dot(vector) {
    this.compareLength(vector);
    return this.val.reduce((a, c, i) => a + c * vector.val[i], 0);
  }
  
  norm() {
    return Math.sqrt(this.val.reduce((a, c) => a + (c ** 2), 0));
  }
  
  equals(vector) {
    if (this.val.length !== vector.val.length) {
      return false;
    }
    
    let result = true;
    this.val.some((v, i) => {
      if (v !== vector.val[i]) {
        result = false;
      }
    });
    
    return result;
  }
  
  toString() {
    return `(${this.val.join(',')})`;
  }
};

var a = new Vector([1, 2, 3]);
var b = new Vector([3, 4, 5]);
var c = new Vector([5, 6, 7, 8]);

a.add(b);      // should return a new Vector([4, 6, 8])
//?
a.subtract(b); // should return a new Vector([-2, -2, -2])
//?
a.dot(b);      // should return 1*3 + 2*4 + 3*5 = 26
//?
a.norm();      // should return sqrt(1^2 + 2^2 + 3^2) = sqrt(14)
//?
// a.add(c);      // throws an error
//?
a.toString();
//?
new Vector([1, 2, 3]).equals(new Vector([3, 4, 5]));
//?
// [ 1, 2, 3 ]
// Vector { val: [ 3, 4, 5 ] }
